PartitionAssignmentsCorrectnessTest Timeout&Failures

```
07:02:10 Tests run: 36, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 269.942 sec - in com.hazelcast.internal.partition.MigrationCorrectnessTest
07:02:10 Running com.hazelcast.internal.partition.PartitionAssignmentsCorrectnessTest
```

```
06:58:38   PartitionAssignmentsCorrectnessTest.testPartitionAssignments_whenNodesStartedTerminated:73->assertPartitionAssignmentsEventually:101->assertPartitionAssignmentsEventually:105->HazelcastTestSupport.assertTrueEventually:871->assertPartitionAssignments:125 Replica 1 is not found in Partition [0]{
06:58:38    0:Address[127.0.0.1]:9976
06:58:38    2:Address[127.0.0.1]:9967
06:58:38    3:Address[127.0.0.1]:9963
06:58:38    4:Address[127.0.0.1]:9961
06:58:38    5:Address[127.0.0.1]:9973
06:58:38 }
```

https://hazelcast-l337.ci.cloudbees.com/view/Official%20Builds/job/Hazelcast-3.x/4969/console
https://hazelcast-l337.ci.cloudbees.com/view/Official%20Builds/job/Hazelcast-3.x-IbmJDK1.7/895/console
https://hazelcast-l337.ci.cloudbees.com/view/Hazelcast/job/Hazelcast-3.x-OracleJDK1.6/com.hazelcast$hazelcast/895/testReport/junit/com.hazelcast.internal.partition/PartitionAssignmentsCorrectnessTest/testPartitionAssignments_whenNodesStartedTerminated_backups_3_nodes_7_/

```
05:42:11 testPartitionAssignments_whenNodesStartedTerminated[backups:3,nodes:7](com.hazelcast.internal.partition.PartitionAssignmentsCorrectnessTest)  Time elapsed: 142.465 sec  <<< FAILURE!
05:42:11 java.lang.AssertionError: Not member: Address[127.0.0.1]:10586
05:42:11 	at org.junit.Assert.fail(Assert.java:88)
05:42:11 	at org.junit.Assert.assertTrue(Assert.java:41)
05:42:11 	at com.hazelcast.internal.partition.PartitionAssignmentsCorrectnessTest.assertPartitionAssignments(PartitionAssignmentsCorrectnessTest.java:126)
```

https://hazelcast-l337.ci.cloudbees.com/view/Hazelcast/job/Hazelcast-3.x-OracleJDK1.6/894/console

--- a/hazelcast/src/main/java/com/hazelcast/internal/partition/impl/MigrationManager.java
+++ b/hazelcast/src/main/java/com/hazelcast/internal/partition/impl/MigrationManager.java
@@ -985,6 +985,25 @@ public class MigrationManager {
                 return;
             }
 
+            Map<Address, Collection<MigrationInfo>> promotions = removeUnknownAddressesAndCollectPromotions();
+            boolean success = promoteBackupsForMissingOwners(promotions);
+
+            partitionServiceLock.lock();
+            try {
+                if (success) {
+                    if (logger.isFinestEnabled()) {
+                        logger.finest("RepartitioningTask scheduled");
+                    }
+                    migrationQueue.add(new RepartitioningTask());
+                } else {
+                    triggerControlTask();
+                }
+            } finally {
+                partitionServiceLock.unlock();
+            }
+        }
+
+        private Map<Address, Collection<MigrationInfo>> removeUnknownAddressesAndCollectPromotions() {
             partitionServiceLock.lock();
             try {
                 Collection<Address> addresses = collectUnknownAddresses();
@@ -992,37 +1011,27 @@ public class MigrationManager {
                     partitionStateManager.removeDeadAddress(address);
                 }
 
-                if (!promoteBackupsForMissingOwners()) {
-                    triggerControlTask();
-                    return;
-                }
+                Map<Address, Collection<MigrationInfo>> promotions = new HashMap<Address, Collection<MigrationInfo>>();
+                for (int partitionId = 0; partitionId < partitionService.getPartitionCount(); partitionId++) {
+                    MigrationInfo migration = createPromotionMigrationIfOwnerIsNull(partitionId);
+                    if (migration == null) {
+                        continue;
+                    }
 
-                if (logger.isFinestEnabled()) {
-                    logger.finest("RepartitioningTask scheduled");
+                    Collection<MigrationInfo> migrations = promotions.get(migration.getDestination());
+                    if (migrations == null) {
+                        migrations = new ArrayList<MigrationInfo>();
+                        promotions.put(migration.getDestination(), migrations);
+                    }
+                    migrations.add(migration);
                 }
-                migrationQueue.add(new RepartitioningTask());
-
+                return promotions;
             } finally {
                 partitionServiceLock.unlock();
             }
         }
 
-        private boolean promoteBackupsForMissingOwners() {
-            Map<Address, Collection<MigrationInfo>> promotions = new HashMap<Address, Collection<MigrationInfo>>();
-            for (int partitionId = 0; partitionId < partitionService.getPartitionCount(); partitionId++) {
-                MigrationInfo migration = createPromotionMigrationIfOwnerIsNull(partitionId);
-                if (migration == null) {
-                    continue;
-                }
-
-                Collection<MigrationInfo> migrations = promotions.get(migration.getDestination());
-                if (migrations == null) {
-                    migrations = new ArrayList<MigrationInfo>();
-                    promotions.put(migration.getDestination(), migrations);
-                }
-                migrations.add(migration);
-            }
-
+        private boolean promoteBackupsForMissingOwners(Map<Address, Collection<MigrationInfo>> promotions) {
             boolean allSucceeded = true;
             for (Map.Entry<Address, Collection<MigrationInfo>> entry : promotions.entrySet()) {
                 Address destination = entry.getKey();
@@ -1037,8 +1046,7 @@ public class MigrationManager {
             boolean success;
             if (node.getThisAddress().equals(destination)) {
                 if (logger.isFinestEnabled()) {
-                    logger.finest("Shortcutting promotions with result: true, "
-                            + "since destination is master.");
+                    logger.finest("Shortcutting promotions with result: true. since destination is master.");
                 }
                 success = true;
             } else {
@@ -1046,28 +1054,31 @@ public class MigrationManager {
                 success = commitMigrationToDestination(destination, migrationsArray);
             }
 
-            for (MigrationInfo migration : migrations) {
+            partitionServiceLock.lock();
+            try {
                 if (success) {
-                    InternalPartitionImpl partition = partitionStateManager.getPartitionImpl(migration.getPartitionId());
+                    for (MigrationInfo migration : migrations) {
+                        InternalPartitionImpl partition = partitionStateManager.getPartitionImpl(migration.getPartitionId());
 
-                    assert partition.getOwnerOrNull() == null : "Owner should be null: " + partition;
-                    assert destination.equals(partition
-                            .getReplicaAddress(migration.getDestinationCurrentReplicaIndex()))
-                            : "Invalid replica! Destination: " + destination
-                                + ", index: " + migration.getDestinationCurrentReplicaIndex() + ", " + partition;
+                        assert partition.getOwnerOrNull() == null : "Owner should be null: " + partition;
+                        assert destination.equals(partition.getReplicaAddress(migration.getDestinationCurrentReplicaIndex()))
+                                : "Invalid replica! Destination: " + destination + ", index: "
+                                        + migration.getDestinationCurrentReplicaIndex() + ", " + partition;
 
-                    partition.swapAddresses(0, migration.getDestinationCurrentReplicaIndex());
-                    addCompletedMigration(migration);
-                    scheduleActiveMigrationFinalization(migration);
+                        partition.swapAddresses(0, migration.getDestinationCurrentReplicaIndex());
+                        addCompletedMigration(migration);
+                        scheduleActiveMigrationFinalization(migration);
+                    }
+                } else {
+                    int delta = migrations.size() + 1;
+                    partitionService.getPartitionStateManager().incrementVersion(delta);
                 }
-            }
-
-            if (!success) {
-                int delta = migrations.size() + 1;
-                partitionService.getPartitionStateManager().incrementVersion(delta);
+            } finally {
+                partitionServiceLock.unlock();
             }
 
             partitionService.syncPartitionRuntimeState();
+
             return success;
         }
 

